src/envs/screws.py

```diff
@@ -0,0 +1,509 @@
+            # variables: [robot, screw to pick up]
+            # params: []
+            "MoveToScrew",
+            self.MoveToScrew_policy,
+            types=[self._gripper_type, self._screw_type])
+        self._MoveToReceptacle: ParameterizedOption = \
+            utils.SingletonParameterizedOption(
+            # variables: [robot, receptacle]
+            # params: []
+            "MoveToReceptacle",
+            self.MoveToReceptacle_policy,
+            types=[self._gripper_type, self._receptacle_type])
+        self._MagnetizeGripper: ParameterizedOption = \
+            utils.SingletonParameterizedOption(
+            # variables: [robot]
+            # params: []
+            "MagnetizeGripper",
+            self.MagnetizeGripper_policy,
+            types=[self._gripper_type])
+        self._DemagnetizeGripper: ParameterizedOption = \
+            utils.SingletonParameterizedOption(
+            # variables: [robot]
+            # params: []
+            "DemagnetizeGripper",
+            self.DemagnetizeGripper_policy,
+            types=[self._gripper_type])
+
+        # Static objects (always exist no matter the settings).
+        self.robot = Object("robby", self._gripper_type)
+        self._receptacle = Object("receptacle", self._receptacle_type)
+
+    def _generate_train_tasks(self) -> List[Task]:
+        return self._get_tasks(num_tasks=CFG.num_train_tasks,
+                               possible_num_screws=self.num_screws_train,
+                               rng=self._train_rng)
+
+    def _generate_test_tasks(self) -> List[Task]:
+        return self._get_tasks(num_tasks=CFG.num_test_tasks,
+                               possible_num_screws=self.num_screws_test,
+                               rng=self._test_rng)
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "screws"
+
+    def render_state(self,
+                     state: State,
+                     task: Task,
+                     action: Optional[Action] = None,
+                     caption: Optional[str] = None) -> List[Image]:
+        fig, ax = plt.subplots(1, 1)
+        plt.xlim([
+            self.rz_x_lb - 2 * self._screw_width,
+            self.rz_x_ub + 2 * self._screw_width
+        ])
+        plt.ylim([self.rz_y_lb, self.rz_y_ub])
+
+        # Draw receptacle
+        receptacle_x = state.get(self._receptacle, "pose_x")
+        receptacle_y = state.get(self._receptacle, "pose_y")
+        receptacle_width = state.get(self._receptacle, "width")
+        plt.plot([
+            receptacle_x - receptacle_width / 2.0,
+            receptacle_x + receptacle_width / 2.0
+        ], [receptacle_y, receptacle_y],
+                 color="green")
+
+        # Draw the gripper
+        gripper_x = state.get(self.robot, "pose_x")
+        gripper_y = state.get(self.robot, "pose_y")
+        gripper_width = state.get(self.robot, "width")
+        # NOTE: this 1.3 factor is just to make the gripper wide enough so that
```

Comment:
the 1.3 thing here is liable to break if we later change the dimensions of stuff. can we accomplish the same thing by changing the conditions for picking so that the entirety of the screw is below the gripper (taking into account the widths of both)?